### DISCUSSION TITLE
backup project on fly.io

### DISCUSSION THREAD

**Discussion by brunowilliang** [pocketbase/pocketbase#1226] [Q&A]
Guys, I have my project on fly.io
How do I generate a backup of it and synchronize with my localhost project?

and as soon as you make the changes perform a new deploy?

![Screen Shot 2022-12-09 at 15 31 53](https://user-images.githubusercontent.com/41432581/206769268-4b2ad30e-affc-41cc-9732-aec7bddcc38e.png)

![Screen Shot 2022-12-09 at 15 35 33](https://user-images.githubusercontent.com/41432581/206769971-a5bbbbad-d96d-4779-81d8-d415826551ec.png)

I was changing the dockerfile by hand and doing a fly deploy, but I needed to synchronize with the offline, so as not to lose the changes

**Comment by ganigeorgiev** [comment]
> How do I generate a backup of it and synchronize with my localhost project?

There are several ways you can download your remote data. One solution would be to use scp and clone the entire `pb_data` directory:

```
# this will register a ssh key with your local agent (if you haven't already)
flyctl ssh issue --agent

# proxies connections to a fly VM through a Wireguard tunnel
flyctl proxy 10022:22

# run in a separate terminal to copy the pb_data directory
scp -r -P 10022 root@localhost:/pb/pb_data  /your/local/path/where/to/copy/pb_data
```

Or use `fly ssh sftp` command as mentioned in https://github.com/pocketbase/pocketbase/discussions/537#discussioncomment-3914729.

---

> I was changing the dockerfile by hand and doing a fly deploy, but I needed to synchronize with the offline, so as not to lose the changes

I'm not understanding this part. Do you mean that you lose your remote data on new deploy? If that's the case make sure that you have mounted the volume in your fly config as shown in https://github.com/pocketbase/pocketbase/discussions/537.

**Reply by brunowilliang** [reply]
I was trying to update pocketbase in this section, is this right?

I change the dockerfile by hand and do a fly deploy...

**Reply by ganigeorgiev** [reply]
Yeah, that's one way to do it. Alternatively you can change the PocketBase version also from your fly config by changing the `build.args` section:

```toml
# optional if you want to change the PocketBase version
[build.args]
  PB_VERSION="0.9.2"
```

**Reply by brunowilliang** [reply]
but I would like to test this version locally first with the data I have on fly.io...
how would i do?

**Reply by ganigeorgiev** [reply]
Sorry, I don't understand what do you mean.

If you run `fly deploy` this will deploy the Dockerfile changes to the remote.

If you want to download your fly data to your local device, run the commands in the comment above (you'll need to adjust only the path where to download the `pb_data`).
Then you can start PocketBase as usual.

If that's not what you are asking, could you clarify what do you want to achieve and/or what commands are you running?

**Reply by brunowilliang** [reply]
that's exactly what I asked, but how do I run my backup locally?

**Reply by brunowilliang** [reply]
I generated the backup and would like to start pocketbase with this backup...

![Screen Shot 2022-12-09 at 16 15 36](https://user-images.githubusercontent.com/41432581/206780197-9304d20d-bb50-45ab-ab48-2a0bc881df9d.png)

**Reply by ganigeorgiev** [reply]
In this case you have 2 options:

1. Download and extract the PocketBase executable on your local device and start it with the `--dir` flag pointing to your directory:
   ```
   ./pocketbase serve --dir="/path/to/your/downloaded/pb_data"
   ```
   
2. Use the Dockerfile and run docker locally. But this is more involving and I don't recommend it if you are not familiar with docker. You'll have to build the image locally, eg. `docker build -t pb .` and then start it with `docker run  -p 8080:8080 -v "/path/to/your/downloaded/pb_data:/pb/pb_data" pb`.

**Reply by brunowilliang** [reply]
omg, perfect, it worked just fine, thanks

**Comment by brunowilliang** [comment]
last question I promise hahaha,
Is there any command where I can update locally?

just paste the cmd ./pocketbase upgrade?

**Reply by ganigeorgiev** [reply]
No. There is a `upgrade` command but it currently is used only for temp data upgrades when migrating from v0.7.x -> v0.8+. In the future this command will be repurposed and eventually we'll add support to auto download the latest github release (it is in the [roadmap](https://github.com/orgs/pocketbase/projects/2)).

**Reply by brunowilliang** [reply]
omg, you guys are wonderful, keep up this project!

**Comment by brunowilliang** [comment]
Hey @ganigeorgiev, I'm going to update my pocketbase to version 0.14.0, I already downloaded the backup.

and then I have a question, if this update goes wrong, crash my application, how can I go back to before?

**Reply by ganigeorgiev** [reply]
Restoring back the old/backup `pb_data` should be enough. You can do that either with `scp` or `rsync`.

**Reply by brunowilliang** [reply]
@ganigeorgiev 
Do you have any code demos? because my application is in production

**Reply by ganigeorgiev** [reply]
The restore is similar to the backup but with reverse arguments. Here is an example upgrade flow:

1. Establish a wireguard tunel:
    ```sh
    # this will register a ssh key with your local agent (if you haven't already)
    flyctl ssh issue --agent

    # proxies connections to a fly VM through a Wireguard tunnel
    flyctl proxy 10022:22
    ```

2. Open a new terminal.

3. Backup the remote `pb_data` to a local directory:
   ```sh
   scp -r -P 10022 root@localhost:/pb/pb_data /your/path/to/backup
   ```

4. Update your fly.toml or Dockerfile to use the latest PocketBase version.

5. Run `flyctl deploy`.

---

**In case something goes wrong:**

1. To restore the previously downloaded backup, execute in the new terminal:
    ```sh
    scp -r -P 10022 /your/path/to/backup/pb_data root@localhost:/pb
    ```

2. Update your fly.toml or Dockerfile to use the latest working PocketBase version.

3. Run `flyctl deploy`.

**Reply by brunowilliang** [reply]
but how would my fly.toml or dockerfile look like?

**Reply by ganigeorgiev** [reply]
If you use the same Dockerfile as in https://github.com/pocketbase/pocketbase/discussions/537, you can either change the `PB_VERSION` instruction OR explicitly define the build arg in your fle.toml:

```toml
[build.args]
  PB_VERSION="0.14.0"
```

**Reply by brunowilliang** [reply]
@ganigeorgiev I ran the command to send the backup and it doesn't work, it sends but doesn't send the data

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what do you mean. I've just tested the above setup locally and it is working fine for me.

Make sure that you are providing the correct paths to your `pb_data` and the parent directory where the files will be copied.

If that doesn't help, please provide as much details as possible, for example the exact commands that you've executed, did you end up using `scp` or `rsync`, what do you mean by _" it sends but doesn't send the data"_, what is your project directory structure, Dockerfile, etc.

**Reply by brunowilliang** [reply]
My steps

```shell
# establish a connection
flyctl proxy 10022:22

# backup download
scp -r -P 10022 root@localhost:/pb/pb_data  /Volumes/Bruno/Dev/Projetos/myfinance/api/pb_data

# restore the backup
scp -r -P 10022 /Volumes/Bruno/Dev/Projetos/myfinance/api/pb_data  root@localhost:/pb/pb_data
```

@ganigeorgiev In my backup I kept some data from a collection, and in mine that is in production I deleted it, and when I run these commands it appears sending the pbdata, but when I refresh my page it doesn't work.

is there any way to use vscode's REMOTE EXPLORER to access my VM and make these changes?

**Reply by ganigeorgiev** [reply]
@Brunowilliang Please check carefully how the paths are defined in  https://github.com/pocketbase/pocketbase/discussions/1226#discussioncomment-5455182.

The first scp command seems that it will download the remote pb_data in a `/Volumes/Bruno/Dev/Projetos/myfinance/api/pb_data/pb_data` dir.

The second scp command will upload the `/Volumes/Bruno/Dev/Projetos/myfinance/api/pb_data` in the remote `/pb/pb_data`, you want only `/pb`.

_Please also note that `scp`, `rsync` and other similar commands usually have a special behavior whether the destination ends with a trailing slash or not._

> is there any way to use vscode's REMOTE EXPLORER to access my VM and make these changes?

I don't use vscode and I'm not familiar with "REMOTE EXPLORER". If you want an interactive sftp/ssh session, you can run something like `flyctl ssh sftp shell` or `flyctl ssh console`.

**Reply by ganigeorgiev** [reply]
Also please note that whatever file changes you decide to do on the remote, you'll have to redeploy or restart the vm (`flyctl restart APP_NAME`) in order to use the new files.

But again, as mentioned in https://github.com/pocketbase/pocketbase/discussions/1226#discussioncomment-5455182, the restore steps is needed only in case something goes wrong with the automated system migrations.

**Reply by brunowilliang** [reply]
@ganigeorgiev I'm not getting it :(
I have two instances of pocketbase on fly.io it connects from ssh but I can't download it, sometimes it works, sometimes it doesn't.

This command was working until yesterday, but today I am no longer able to restore the backup
